src/Bicep.Core/TypeSystem/TypeAssignmentVisitor.cs

```diff
@@ -11,55 +11,64 @@
 using Bicep.Core.Resources;
 using Bicep.Core.SemanticModel;
 using Bicep.Core.Syntax;
+using Bicep.Core.Syntax.Visitors;
 
 namespace Bicep.Core.TypeSystem
 {
     public sealed class TypeAssignmentVisitor : SyntaxVisitor
     {
+        public class TypeAssignment
+        {
+            public TypeAssignment(ITypeReference reference)
+                : this(reference, Enumerable.Empty<Diagnostic>())
+            {
+            }
+
+            public TypeAssignment(ITypeReference reference, IEnumerable<Diagnostic> diagnostics)
+            {
+                Reference = reference;
+                Diagnostics = diagnostics;
+            }
+
+            public ITypeReference Reference { get; }
+
+            public IEnumerable<Diagnostic> Diagnostics { get; }
+        }
+
         private readonly ResourceTypeRegistrar resourceTypeRegistrar;
+        private readonly TypeManager typeManager;
         private readonly IReadOnlyDictionary<SyntaxBase, Symbol> bindings;
-
         private readonly IReadOnlyDictionary<SyntaxBase, ImmutableArray<DeclaredSymbol>> cyclesBySyntax;
+        private IDictionary<SyntaxBase, TypeAssignment> assignedTypes;
 
-        private readonly IDictionary<SyntaxBase, ITypeReference> assignedTypes;
-
-        public TypeAssignmentVisitor(ResourceTypeRegistrar resourceTypeRegistrar, IReadOnlyDictionary<SyntaxBase, Symbol> bindings, IReadOnlyDictionary<SyntaxBase, ImmutableArray<DeclaredSymbol>> cyclesBySyntax)
+        public TypeAssignmentVisitor(ResourceTypeRegistrar resourceTypeRegistrar, TypeManager typeManager, IReadOnlyDictionary<SyntaxBase, Symbol> bindings, IReadOnlyDictionary<SyntaxBase, ImmutableArray<DeclaredSymbol>> cyclesBySyntax)
```

Comment:
Things I don't really like about this solution:
1. Recursion happens through the `typeManager` instance - this class calls `typeManager.GetTypeInfo(..)` which calls back into this class etc. This is needed to provide a reference to `TypeValidator` to use.
1. This class is getting larger and larger.